src/Client.php

```diff
@@ -0,0 +1,197 @@
+
+            if (is_array($value)) {
+                $output += $this->flatten($value, $finalKey); // @todo: handle name collision here if needed
+            } else {
+                $output[$finalKey] = $value;
+            }
+        }
+
+        return $output;
+    }
+  /**
+     * Private static method to process the data before making the request
+     * @param array<string, string> $headers
+     * @param array<string>|array<string, mixed> $body
+     * @param array<string, mixed> $query
+     * @param string $method
+     * @param string $url
+     */
+    private function processData(
+        array &$headers,
+        array &$body,
+        array $query,
+        string &$method,
+        string &$url
+    ): void {
+        if(!$method) { // if method is not set, set it to GET by default
+            $method = self::METHOD_GET;
+        } else { // else convert the method to uppercase
+            $method = strtoupper($method);
+        }
+        // if there are no headers but a body set header to json by default
+        if(!isset($headers['content-type']) && is_array($body) && count($body) > 0) {
+            $headers['content-type'] = 'application/json';
+        }
+        if(isset($headers['content-type'])) {
+            switch($headers['content-type']) { // Convert the body to the appropriate format
+                case 'application/json':
+                    $body = json_encode($body);
+                    break;
+                case 'application/x-www-form-urlencoded':
+                    $body = $this->flatten($body);
+                    break;
+                case 'multipart/form-data':
+                    $body = $this->flatten($body);
+                    break;
+                case 'application/graphql':
+                    $body = $body[0];
+                    break;
+                default:
+            }
+        }
+        if(!is_array($headers)) {
+            $headers = [];
+        }
+        foreach ($headers as $i => $header) { // convert headers to appropriate format
+            $headers[] = $i . ':' . $header;
+            unset($headers[$i]);
+        }
+        if($query) {  // if the request has a query string, append it to the request URI
+            $url .= '?' . http_build_query($query);
+        }
+    }
+  /**
+   * This method is used to make a request to the server
+   * @param string $url
+   * @param array<string, string> $headers
+   * @param string $method
+   * @param array<string>|array<string, mixed> $body
+   * @param array<string, mixed> $query
+   * @return Response
+   */
+    private function fetchHelper(
+        string $url,
+        array $headers,
+        string $method,
+        array $body,
+        array $query
+    ): Response {
+        // Process the data before making the request
+        $this->processData(
+            $headers,
+            $body,
+            $query,
+            $method,
+            $url
+        );
```

Comment:
this function is a little bit hard to understand, since many values are changed by reference. I think it might make more sense to break the processData function up into smaller functions.

For example individual function to handle. We could also just leave all the logic in a big single fetchHelper function.